### PR TITLE
Alfoa/opt features

### LINKED ISSUE #111: GradientBasedOptimizer gradient feature space
--------
Issue Description - FEATURE
--------
##### Which new enhancement should be added?

The normalization of the feature space in the gradient-based optimizers should be done not only for the numerical stability but also for the gradient computation

##### Why?

This enhancement is going to make the opt path much more resilient and will facilitate the optimization parameter settings

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [X] 1. Is it tagged with a type: defect or improvement? ***improvement***
- [X] 2. Is it tagged with a priority: critical, normal or minor? ***critical***
- [X] 3. If it will impact requirements or requirements tests, is it tagged with requirements? ***n/a***
- [X] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users. ***n/a***
- [X] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.) ***yes***

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed? ***not a defect***
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.) ***not a defect***
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)? ***not a defect***
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)? ***not a defect***
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided? ***#207t***

### PR DESCRIPTION
**Pull request by alfoa**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)

Closes #111 

##### What are the significant changes in functionality due to this change request?

This PR addresses issue #111. The normalization of the gradient and feature space is always done.
In addition, it adds an initial implementation of an adaptive stepping approach (line search using a surrogate model). This approach is not optimal yet but gives better performances than before.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

**Comment by alfoa** [comment]
@PaulTalbot-INL It is ready to be reviewed.

File: framework/Optimizers/SPSA.py [hunk 1]
```diff
@@ -364,47 +389,47 @@ def _generateVarsUpdateConstrained(self,ak,gradient,varK):
...
       self.raiseAnError(ValueError, 'The innerBisectionThreshold shall be greater than 0 and less than 1')
-    fracLowerLimit = 1e-2
     bounds = [0, 1.0]
     tempVarNew = {}
     frac = 0.5
     while np.absolute(bounds[1]-bounds[0]) >= innerBisectionThreshold:
-      for var in self.optVars:
-        tempVarNew[var] = copy.copy(varK[var]-gain*vector[var]*1.0*frac)
+      for index, var in enumerate(self.optVars):
+        tempVarNew[var] = copy.copy(varK[var]-gain[index]*vector[var]*1.0*frac)
```
Comment (PaulTalbot-INL) [inline framework/Optimizers/SPSA.py]: is the `*1.0` necessary with the `*frac` included?

File: framework/Optimizers/SPSA.py [hunk 2]
```diff
@@ -441,23 +466,61 @@ def _computeGainSequenceCk(self,paramDict,iterNum):
...
+          Method that just interface the computes the approximate derivatives using the surrogate
+          @ In, x, numpy.array, coordinate where to evaluate f'
+          @ Out, f, numpy.array, partial derivatives
+        """
+        return scipy.optimize.approx_fprime(x, f, self._computeGainSequenceCk(self.paramDict,self.counter['varsUpdate'][traj]+1))
+
+      xK             = np.asarray([self.optVarsHist[traj][iterNum-1][key] for key in self.optVars])
+      xKPrevious     = np.asarray([self.optVarsHist[traj][iterNum-2][key] for key in self.optVars])
+      #xK             = np.asarray([self.denormalizeData(self.optVarsHist[traj][iterNum-1])[key] for key in self.optVars])
+      #xKPrevious     = np.asarray([self.denormalizeData(self.optVarsHist[traj][iterNum-2])[key] for key in self.optVars])
```
Comment (PaulTalbot-INL) [inline framework/Optimizers/SPSA.py]: A few more extraneous comment lines

File: tests/framework/Optimizers/GradientBasedOptimizers/t.xml [hunk 3]
```diff
@@ -1,25 +1,20 @@
...
+        This test is aimed to check the capability of RAVEN to input constant variables values. This
+        test is valid to demonstrate this capability for any Optimizer in RAVEN.
     </description>
   </TestInfo>
 
   <!-- actual RAVEN input -->
-
   <RunInfo>
-    <WorkingDir>maxRangeWithBoundaryConstraintsNormGradient</WorkingDir>
+    <WorkingDir>/Users/alfoa/projects/raven_github/raven/tests/framework/Optimizers/GradientBasedOptimizers/maxRangeWithBoundaryConstraintsAndConstants</WorkingDir>
```
Comment (PaulTalbot-INL) [inline tests/framework/Optimizers/GradientBasedOptimizers/t.xml]: hard-coded user directory might be a problem

File: tests/framework/Optimizers/GradientBasedOptimizers/t.xml [hunk 4]
```diff
@@ -1,25 +1,20 @@
 <?xml version="1.0" ?>
```
Comment (PaulTalbot-INL) [inline tests/framework/Optimizers/GradientBasedOptimizers/t.xml]: why is this test named "t" and replacing the max_range_projectile_norm_grad test?

File: tests/framework/Optimizers/GradientBasedOptimizers/test_spsaWConstraint.xml [hunk 5]
```diff
@@ -61,30 +61,30 @@
   <Optimizers>
     <SPSA name="opt_smp">
       <initialization>
-        <limit>1000</limit>
+        <limit>10000</limit>
```
Comment (PaulTalbot-INL) [inline tests/framework/Optimizers/GradientBasedOptimizers/test_spsaWConstraint.xml]: How long does the 10k samples take?  Or is this just a high limit so we hit IterationLimit instead?

**Pull request merged**